pallets/pooled-staking/src/pools.rs

```diff
@@ -0,0 +1,393 @@
+                Stake(Pools::<T>::get(
+                    candidate,
+                    &PoolsKey::$hold {
+                        delegator: delegator.clone(),
+                    },
+                ))
+            }
+
+            fn set_shares(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Shares<T>) {
+                Pools::<T>::set(
+                    candidate,
+                    &PoolsKey::$shares {
+                        delegator: delegator.clone(),
+                    },
+                    value.0,
+                )
+            }
+
+            fn set_shares_supply(candidate: &Candidate<T>, value: Shares<T>) {
+                Pools::<T>::set(candidate, &PoolsKey::$supply, value.0)
+            }
+
+            fn set_total_staked(candidate: &Candidate<T>, value: Stake<T>) {
+                Pools::<T>::set(candidate, &PoolsKey::$total, value.0)
+            }
+
+            fn set_hold(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Stake<T>) {
+                Pools::<T>::set(
+                    candidate,
+                    &PoolsKey::$hold {
+                        delegator: delegator.clone(),
+                    },
+                    value.0,
+                )
+            }
+
+            fn initial_share_value() -> Stake<T> {
+                Stake($init)
+            }
+        }
+    };
+}
+
+impl_pool!(
+    Joining,
+    JoiningShares,
+    JoiningSharesSupply,
+    JoiningSharesTotalStaked,
+    JoiningSharesHeldStake,
+    T::InitialJoiningShareValue::get(),
+);
+
+impl_pool!(
+    AutoCompounding,
+    AutoCompoundingShares,
+    AutoCompoundingSharesSupply,
+    AutoCompoundingSharesTotalStaked,
+    AutoCompoundingSharesHeldStake,
+    T::InitialAutoCompoundingShareValue::get(),
+);
+
+impl_pool!(
+    ManualRewards,
+    ManualRewardsShares,
+    ManualRewardsSharesSupply,
+    ManualRewardsSharesTotalStaked,
+    ManualRewardsSharesHeldStake,
+    T::InitialManualClaimShareValue::get(),
+);
+
+impl_pool!(
+    Leaving,
+    LeavingShares,
+    LeavingSharesSupply,
+    LeavingSharesTotalStaked,
+    LeavingSharesHeldStake,
+    T::InitialLeavingShareValue::get(),
+);
+
+impl<T: Config> ManualRewards<T> {
+    pub fn pending_rewards(
+        candidate: &Candidate<T>,
+        delegator: &Delegator<T>,
+    ) -> Result<Stake<T>, Error<T>> {
+        let shares = Self::shares(candidate, delegator);
+
+        if shares.0.is_zero() {
+            return Ok(Stake(0u32.into()));
+        }
+
+        let counter = Pools::<T>::get(candidate, &PoolsKey::ManualRewardsCounter);
+        let checkpoint = Pools::<T>::get(
+            candidate,
+            &PoolsKey::ManualRewardsCheckpoint {
+                delegator: delegator.clone(),
+            },
+        );
+
+        // TODO: Should be safe to wrap around.
+        let diff = counter.err_sub(&checkpoint)?;
+        Ok(Stake(diff.err_mul(&shares.0)?))
+    }
+
+    pub fn claim_rewards(
```

Comment:
This function and `pending_rewards` are unused, so maybe remove them until they are needed? That way we can review them better later.

Reply:
This part should be reviewed now as they deal with the counter/checkpoint system. If there is an issue in there it will require massive migrations to fix the checkpoints for all delegators. I'll write tests to ensure they work properly.